install_sct

```diff
@@ -641,19 +641,16 @@ fi
 
 
 ## Check if installed TF/Keras requirements throw instruction set errors
-if python3 -c "import keras; print(keras.__version__)" 2>&1 | grep -q 'The TensorFlow library was compiled to use'; then
+if python3 -c "import keras; print(keras.__version__)" 2>&1 | grep -q 'The TensorFlow library was compiled to use\|Illegal instruction'; then
```

Comment:
This conditional [doesn't seem to trigger](https://neuropoly.slack.com/archives/CAW9X21D5/p1636243175074100) on @mathieuboudreau's M1 Mac, and I can't figure out why.

This is the testing I did for this conditional back in #3361:

```console
joshua@tadpole:~$ qemu-x86_64 --cpu Westmere-v1 /home/joshua/repos/spinalcordtoolbox_test-python-3.7-tensorflow-1.15/python/envs/venv_sct/bin/python3 -c "import keras; print(keras.__version__)" 2>&1 |       grep 'The TensorFlow library was compiled to use\|Illegal instruction'
2021-11-07 09:56:12.161788: F tensorflow/core/platform/cpu_feature_guard.cc:37] The TensorFlow library was compiled to use AVX instructions, but these aren't available on your machine.
```

My best guess is that  `python3 -c "import keras; print(keras.__version__)"` behaves differently when called on an M1 Mac compared to calling it via `qemu-x86_64`. Maybe it triggers some sort of crash that prevents the output from being captured?

Reply:
@kousu Is there a better way to perform this check? (Shot in the dark, but... would using a subprocess help here? I'm not really sure how to isolate the Python call.)

Reply:
@mathieuboudreau: Could you share the output when you run the following commands on your M1 Mac?

```
# Without subshell
$SCT_DIR/python/envs/venv_sct/bin/python3 -c "import keras; print(keras.__version__)" 2>&1 | grep 'The TensorFlow library was compiled to use\|Illegal instruction' || echo "Illegal instruction not matched" 

# With subshell (not sure if this makes any difference)
echo $($SCT_DIR/python/envs/venv_sct/bin/python3 -c "import keras; print(keras.__version__)" 2>&1 | grep 'The TensorFlow library was compiled to use\|Illegal instruction' || echo "Illegal instruction not matched")
```

Reply:
```shell
(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ # Without subshell
(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ $SCT_DIR/python/envs/venv_sct/bin/python3 -c "import keras; print(keras.__version__)" 2>&1 | grep 'The TensorFlow library was compiled to use\|Illegal instruction' || echo "Illegal instruction not matched" 

Illegal instruction not matched
(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ 
(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ # With subshell (not sure if this makes any difference)
(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ echo $($SCT_DIR/python/envs/venv_sct/bin/python3 -c "import keras; print(keras.__version__)" 2>&1 | grep 'The TensorFlow library was compiled to use\|Illegal instruction' || echo "Illegal instruction not matched")
Illegal instruction not matched
(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ 
```

Reply:
I think the problem is you need to use `egrep` (or `grep -E`). `|` is a regexp character and `grep` doesn't do those by default.

You could also split it up like

```
if ( (python3 -c "import keras; print(keras.__version__)" 2>&1 | grep -q 'The TensorFlow library was compiled to use) || (python3 -c "import keras; print(keras.__version__)" 2>&1 | grep -q 'Illegal instruction') ); then 
```

Reply:
> I think the problem is you need to use `egrep` (or `grep -E`). `|` is a regexp character and `grep` doesn't do those by default.

I had been following [this blog post](https://www.thegeekstuff.com/2011/10/grep-or-and-not-operators/) which suggests that escaping `|` (i.e. `\|`) allows it to be used with regular ol' `grep`.

Reply:
@joshuacwnewton let me know when you want me to test it again!

Reply:
macs might have BSD grep, not Linux grep? It's a different dialect.

https://ss64.com/osx/grep.html says macOS has `grep -E` so..that should work?

Reply:
I'm willing to schedule a videocall if you want to tinker with it.

Reply:
Aside from the grep issues, we can still quickly test the wheel itself manually for now. :smile: 

@mathieuboudreau Could you try the following steps and see if it fixes the issue?

```bash
source ${SCT_DIR}/python/etc/profile.d/conda.sh
conda activate venv_sct
pip uninstall tensorflow
pip install https://github.com/spinalcordtoolbox/docker-tensorflow-builder/releases/download/v1.15.5-py3.7/tensorflow-1.15.0-cp37-cp37m-macos_arm64.whl
sct_check_dependencies
```

If all goes well, you shouldn't need to reinstall SCT from scratch -- you should just be able to immediately run `sct_check_dependencies` without error.

Reply:
@joshuacwnewton I get an error, see below:

```shell

(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ source ${SCT_DIR}/python/etc/profile.d/conda.sh
(base) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ conda activate venv_sct
(venv_sct) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ pip uninstall tensorflow
Found existing installation: tensorflow 1.15.5
Uninstalling tensorflow-1.15.5:
  Would remove:
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/estimator_ckpt_converter
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/freeze_graph
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/saved_model_cli
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/tensorboard
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/tf_upgrade_v2
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/tflite_convert
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/toco
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/bin/toco_from_protos
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/tensorflow-1.15.5.dist-info/*
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/tensorflow/*
    /Users/sct_m1/spinalcordtoolbox/python/envs/venv_sct/lib/python3.7/site-packages/tensorflow_core/*
Proceed (y/n)? y
  Successfully uninstalled tensorflow-1.15.5
(venv_sct) Mathieus-MacBook-Pro:spinalcordtoolbox sct_m1$ pip install https://github.com/spinalcordtoolbox/docker-tensorflow-builder/releases/download/v1.15.5-py3.7/tensorflow-1.15.0-cp37-cp37m-macos_arm64.whl
ERROR: tensorflow-1.15.0-cp37-cp37m-macos_arm64.whl is not a supported wheel on this platform.

```

Reply:
:( drat. That seemed so promising.

That thread *said* they'd got it working on an M1, right?

Reply:
Which thread? Is it possible that I need to be on the most recent OS for it to work?

Reply:
I'm still on Big Sur.

Reply:
Looks good! 🚀
<img width="1440" alt="Screen Shot 2021-11-09 at 11 00 58 AM" src="https://user-images.githubusercontent.com/1421029/140948602-23fe4539-fb39-48d6-85fb-ecfaae290780.png">

Now I'm hitting an FSLeyes error, it seems it hasn't been installed. Was I supposed to install it first? I know there's a separate hack needed to get FSLeyes working on recent Mac OS's.

Reply:
That's not an error, thankfully. FSLeyes is an optional dependency.

But, maybe the wording needs to be tweaked to make it a little friendlier. It does look like an error, doesn't it? :sweat_smile:

Reply:
OK, I think the reason we couldn't catch the error is because we're piping the output of `python3 -c` into `grep`. But, the `python3` command naturally fails (due to the illegal instruction), _and_ we have `pipefail` set in `install_sct`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/7fd2ea718751dd858840c3823c0830a910d9777c/install_sct#L36

So, I think the command we actually want to use is:

```bash
if grep -q 'The TensorFlow library was compiled to use\|Illegal instruction' <<< "$(python3 -c "import keras; print(keras.__version__)" 2>&1)"; then
```

Using a subshell + a [here string](https://tldp.org/LDP/abs/html/x17837.html) should let us more reliably catch the output of the check... I hope? But, @kousu is a more skilled bash wizard than I, so... maybe I'm missing something here. :P

----

@mathieuboudreau Sorry to bother again, but could you try installing SCT from this branch and see if you run into the error again?

Reply:
Scratch that -- @valosekj gave this a try, and this solution still didn't catch the error. :thinking: 

![image](https://user-images.githubusercontent.com/16181459/143497108-c9448310-6ee9-4f29-bb39-e9585bb4f82d.png)

Reply:
Sorry, I missed this in my emails! Feel free to ping me on Slack next time if it's urgent!

Reply:
I think this has been resolved now by 978d1cd. :slightly_smiling_face:

Reply:
Want a fresh test?

Reply:
That would be wonderful, thank you!

Reply:
Looks good on my end! 

Flawless victory <3 
<img width="889" alt="Screen Shot 2021-11-30 at 2 48 44 PM" src="https://user-images.githubusercontent.com/1421029/144109274-40e498de-4bb6-452e-8d44-c47b8c8a1cee.png">

Reply:
Installation finished successfully for me as well! On both M1 MacBook Air and M1 Mac mini. Thanks a lot @joshuacwnewton!